### ISSUE TITLE
Rule: find unused private top level properties, constants or functions

### ISSUE THREAD

**Issue by vanniktech** [arturbosch/detekt#948]
The following does not yield a UnusedMemberDeclaration warning:

```kotlin
fun foo(map: Array<IntArray>) {
  for (i in 0 until map.size) {
    for (j in 0 until map[i].size) {
      map[i][i] += 1
    }
  }
}
```

**Comment by arturbosch** [comment]
maybe because it is a top level function and `UnusedMemberDeclaration` only detects members?
So this would be a feature request? :D

**Comment by vanniktech** [comment]
Huh indeed.

**Comment by vanniktech** [comment]
Wrapping inside a class does trigger the rule:

```kotlin
class Foo {
  fun foo(map: Array<IntArray>) {
    for (i in 0 until map.size) {
      for (j in 0 until map[i].size) {
        map[i][i] += 1
      }
    }
  }
}
```

**Comment by marianosimone** [comment]
I'd love jump in and work on something like this, but I'm a bit lost: I can't even find `UnusedMemberDeclaration`

**Comment by vanniktech** [comment]
https://github.com/arturbosch/detekt/blob/ad1808b899aacd842e434704aa451145317cf211/detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/UnusedPrivateMember.kt

**Comment by schalkms** [comment]
@marianosimone contributions are always welcome 👍 
As mentioned `UnusedPrivateMember.kt` is the way to go. I would add this feature in a new rule. Some of the functionality of `UnusedPrivateMember` might even be reused for the new rule.

**Comment by marianosimone** [comment]
I've been working on this, based on what `UnusedPrivateMember` does, and I wanted to get some feedback on what I'm thinking about before creating a PR:

`UnusedPrivateMember` seems to be doing quite a bit, and not everything is related to the rest. From the docs:

>  Reports unused private properties, function parameters and functions.

So far, this detects:
- Unused private top-level functions
- Unused private member functions
- Unused function parameters (only for non-top-level functions)
- Unused private properties (non-top-level)
- Unused private constants (non-top-level)
- Unused variable declaration inside functions (only for non-top-level functions)

The way I see it, detecting "Unused private top-level functions" is the odd one there: everything else only works when dealing with non-top-level.

Also, IMO, detecting "unused parameters" is out of place, as there's no concept of private vs non private here.

A proposal:
- Create a `UnusedFunctionParameter` rule, that only looks at parameters (gracefully handling `override` cases, in which the parameter needs to be declared, even if not used). The logic is almost all already present in `UnusedPrivateMember`, I'd just need to move it out. I'm still 
- Add unused private inner class detection to `UnusedPrivateMember`, as a consequence of that and the previous point, `UnusedPrivateMember` documentation should be: ">  Reports unused private properties, member functions and inner classes." _NOTE: still not sure if I'd rather have this or an independent rule that looks for unused private classes, being them either top-level or inner_
- Create `UnusedPrivateTopLevelDeclaration` rule, that basically mimics `UnusedPrivateMember`, but for top-level declarations.

An alternative is accepting the fact that `UnusedPrivateMember` already deals with top-level functions somehow, and make it do even more (`UnusedPrivateDeclaration` could be a better name at this point, but we'd have to keep the existing rule for backwards-compatibility (?)):
- Unused private functions
- Unused function parameters
- Unused private properties
- Unused private constants
- Unused variable declaration inside functions

I DO already have code for this alternative, but basically wanted to know what people think about what the best option would be, before I iron details out and upload for review ;)

**Comment by schalkms** [comment]
@marianosimone 
Personally, I think both approaches are fine, even though I prefer the second one with well defined (inner) classes for those violations. We wouldn’t need to change much.
Big thanks and I am looking forward to reviewing your implementation!
